### PR TITLE
Matrix-free operator cleanup 

### LINKED ISSUE #3295: linear_operator and LA::distributed::Vector<> with MatrixFree
looks like the two can not be combined at the moment due to specifics of initialization by ``MatrixFree::initialize_dof_vector()` instead of `vector.initialize(locally_owned_dofs, mpi_communicator`:

```
The violated condition was:
   dst.partitioners_are_globally_compatible(*data->get_dof_info(0).vector_partitioner)
```

Can we overcome this limitation somehow?

### LINKED ISSUE #3343: Make DiagonalMatrix member of MatrixFreeOperators::Base shared_ptr
If we have an `std_cxx11::shared_ptr<DiagonalPreconditioner<VectorType> >` in the `MatrixFreeOperators::Base` class and a return function for the shared pointer, we can share the storage inside the class with the outside world, e.g. in step-37 according to #3329.

### LINKED ISSUE #3363: Static analysis: dealii-git/include/deal.II/lac/precondition.h
```
dealii-git/include/deal.II/lac/precondition.h	2082	warn	V560 A part of conditional expression is always true.
dealii-git/include/deal.II/lac/precondition.h	2084	warn	V560 A part of conditional expression is always true.
dealii-git/include/deal.II/lac/precondition.h	2081	warn	V560 A part of conditional expression is always true.
dealii-git/include/deal.II/lac/precondition.h	1955	warn	V730 Not all members of a class are initialized inside the constructor. Consider inspecting: theta, delta.
dealii-git/include/deal.II/lac/precondition.h	1934	err	V730 Not all members of a class are initialized inside the constructor. Consider inspecting: matrix_diagonal_inverse.
```

We should address these warnings and errors from the static analysis tool PVS. In response to #3342.

### LINKED ISSUE #3447: Deprecated warnings in MatrixFree::reinit() without IndexSets
`reinit()` functions with `locally_owned_dofs` were deprecated recently, however their alternatives still use deprecated functions inside:

```
/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/dealii-develop-n63cfzji36xylrgulz7s5xsdisphv7xv/include/deal.II/matrix_free/matrix_free.h:368:8: note: 'reinit' has been explicitly marked deprecated here
  void reinit (const Mapping<dim>                          &mapping,
```
For example 
```
template <int dim, typename Number>
template <typename DoFHandlerType, typename QuadratureType>
void MatrixFree<dim,Number>::
reinit(const Mapping<dim>                                   &mapping,
       const std::vector<const DoFHandlerType *>            &dof_handler,
       const std::vector<const ConstraintMatrix *>          &constraint,
       const std::vector<QuadratureType>                    &quad,
       const typename MatrixFree<dim,Number>::AdditionalData additional_data)
{
  std::vector<IndexSet> locally_owned_set =
    internal::MatrixFree::extract_locally_owned_index_sets
    (dof_handler, additional_data.level_mg_handler);
  reinit(mapping, dof_handler,constraint,locally_owned_set,
         quad, additional_data);
}
```

### PR DESCRIPTION
**Pull request by kronbichler**
Fixes #3447.
Fixes #3343.
Fixes #3295.
Fixes #3363.

The first few commits in this PR are taken from #3329. I want to provide those improvements more quickly because it will take me a bit more time to finish the tutorial.

This PR also cleans up a small mistake from the PR #3294 where I used the rhs vector rather than the vector consisting of ones. This can sometimes lead to no convergence and bad estimation of eigenvalues, which is why I had to adjust the output of `parallel_multigrid.output`

### CODE REVIEW

File: include/deal.II/lac/precondition.h [hunk 1]
```diff
@@ -856,6 +856,12 @@ class PreconditionPSOR : public PreconditionRelaxation<MatrixType>
  * be used to determine the number of iteration, see e.g. section 5.1 of
  * R. S. Varga, Matrix iterative analysis, 2nd ed., Springer, 2009.
  *
+ * In order to use Chebyshev as a solver, set the degree to
+ * numbers::invalid_unsigned_int to force the automatic computation of the
+ * number of iterations needed to reach a given target tolerance. In this
+ * case, the target tolerance is read from the variable smoothing_range (it
```
Comment (bangerth) [inline include/deal.II/lac/precondition.h]: If you qualify the variable name with the class name, doxygen will automatically link to it.

File: include/deal.II/lac/precondition.h [hunk 2]
```diff
@@ -911,19 +917,24 @@ class PreconditionChebyshev : public Subscriptor
...
      * This sets the range between the largest eigenvalue in the matrix and
-     * the smallest eigenvalue to be treated. If the parameter is zero, an
-     * estimate for the largest and for the smallest eigenvalue will be
-     * calculated internally. Otherwise, the Chebyshev polynomial will act in
-     * the interval $[\lambda_\mathrm{max}/ \tt{smoothing\_range},
-     * \lambda_\mathrm{max}]$, where $\lambda_\mathrm{max}$ is an estimate of
-     * the maximum eigenvalue of the matrix. A choice of
-     * <tt>smoothing_range</tt> between 5 and 20 is useful in case the
-     * preconditioner is used as a smoother in multigrid.
+     * the smallest eigenvalue to be treated. If the parameter to a number
```
Comment (bangerth) [inline include/deal.II/lac/precondition.h]: If the parameter IS SET to a number?

File: include/deal.II/lac/precondition.h [hunk 3]
```diff
@@ -2073,6 +2081,21 @@ PreconditionChebyshev<MatrixType,VectorType,PreconditionerType>::estimate_eigenv
   const double alpha = (data.smoothing_range > 1. ?
                         max_eigenvalue / data.smoothing_range :
                         std::min(0.9*max_eigenvalue, min_eigenvalue));
+
+  // in case the user set the degree to invalid unsigned int, we have to
+  // determine the number of necessary iterations from the Chebyshev error
+  // estimate, given the target tolerance specified by smoothing_range. This
+  // estimate is based on the error formula given in section 5.1 of
+  // R. S. Varga, Matrix iterative analysis, 2nd ed., Springer, 2009
```
Comment (davydden) [inline include/deal.II/lac/precondition.h]: is this reference mentioned in class documentation? If not, i would add it somewhere as a bibtex entry.

Comment (davydden) [inline include/deal.II/lac/precondition.h]: p.s. of course, you don't have to remove this in-code comment ;-)

File: include/deal.II/matrix_free/operators.h [hunk 4]
```diff
@@ -202,7 +203,7 @@ namespace MatrixFreeOperators
     /**
      * A vector to store inverse of diagonal elements.
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: while there, could you adjust the documentation?

File: include/deal.II/matrix_free/operators.h [hunk 5]
```diff
@@ -678,13 +686,18 @@ namespace MatrixFreeOperators
...
-              const MGConstrainedDoFs &mg_constrained_dofs,
-              const unsigned int level)
+              const MGConstrainedDoFs      &mg_constrained_dofs,
+              const unsigned int            level)
   {
     AssertThrow (level != numbers::invalid_unsigned_int,
                  ExcMessage("level is not set"));
+    if (data_.n_macro_cells() > 0)
+      {
+        AssertDimension(static_cast<int>(level),
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: essentially, it makes sure that the provided `data` object corresponds to the provided `level`? Are there cases when `data_.n_macro_cells()` is not yet positive? I guess not. If so, we may as well remove the `level `from the arguments and take it from this assertion.

File: include/deal.II/matrix_free/operators.h [hunk 6]
```diff
@@ -229,6 +230,12 @@ namespace MatrixFreeOperators
     void mult_add (LinearAlgebra::distributed::Vector<Number> &dst,
                    const LinearAlgebra::distributed::Vector<Number> &src,
                    const bool transpose) const;
+
+    /**
+     * Adjust the ghost range of the vectors to the storage requirements of
+     * the underlying MatrixFree class.
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: maybe give a hint in documentation where that could be needed/used (i see it is in `mult_add`), in what circumstances?

File: include/deal.II/matrix_free/operators.h [hunk 7]
```diff
@@ -935,23 +978,23 @@ namespace MatrixFreeOperators
...
 
-    const unsigned int local_size = inverse_diagonal_entries.local_size();
+    const unsigned int local_size = inverse_diagonal_vector.local_size();
     for (unsigned int i=0; i<local_size; ++i)
-      inverse_diagonal_entries.local_element(i)
-        =1./inverse_diagonal_entries.local_element(i);
+      inverse_diagonal_vector.local_element(i)
+        =1./inverse_diagonal_vector.local_element(i);
 
-    inverse_diagonal_entries.compress(VectorOperation::insert);
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: i was also curious whether this is needed or not. :+1:

File: include/deal.II/matrix_free/operators.h [hunk 8]
```diff
@@ -1043,22 +1086,24 @@ namespace MatrixFreeOperators
...
     unsigned int dummy = 0;
-    LinearAlgebra::distributed::Vector<Number> &inverse_diagonal_entries = Base<dim,Number>::inverse_diagonal_entries.get_vector();
-    this->initialize_dof_vector(inverse_diagonal_entries);
-    Base<dim,Number>::
-    data->cell_loop (&LaplaceOperator::local_diagonal_cell,
-                     this, inverse_diagonal_entries, dummy);
+    this->inverse_diagonal_entries.
+    reset(new DiagonalMatrix<LinearAlgebra::distributed::Vector<Number> >());
+    LinearAlgebra::distributed::Vector<Number> &inverse_diagonal_vector = this->inverse_diagonal_entries->get_vector();
+    LinearAlgebra::distributed::Vector<Number> ones;
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: presumably a copy-paste error, there is no need for `ones` here.

File: include/deal.II/matrix_free/operators.h (hunk 5 above)
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: On a second thought, it is probablybetter to keep 'level' argument as it makes it more clear that it is intended to be used with GMG. Also n_macro_cells could probably be zero in MPI cases for some cores on some levels. Bottom line -- never mind changing anytging here, i think it is all good.

Comment (kronbichler) [inline include/deal.II/matrix_free/operators.h]: I also thought about removing the level but decided that it is less clear without it. So we seem to agree on this topic! :+1:

File: include/deal.II/lac/precondition.h (hunk 2 above)
Comment (kronbichler) [inline include/deal.II/lac/precondition.h]: Of course.

**Comment by kronbichler** [comment]
I think I addressed all comments.

**Comment by kronbichler** [comment]
This now has both tests renamed.

**Pull request merged**